### ISSUE TITLE
Running logstash 6.x in Docker Swarm creates nonconfigurable container

### ISSUE THREAD

**Issue by vlindhol** [elastic/logstash-docker#88]
I have the following simple `docker-compose.yml` file:

```yml
version: "3"

services:
  logstash:
      image: docker.elastic.co/logstash/logstash:6.2.4
      volumes:
        - ./logstash/config/:/usr/share/logstash/config/
        - ./logstash/pipeline/:/usr/share/logstash/pipeline/
      deploy:
        mode: replicated
        replicas: 1
```

where the pipeline is configured with a TCP input and elasticsearch output. Logstash 6.x in Docker fails to read the configuration when run in a Docker Swarm, i.e. with the command `docker swarm init; docker stack deploy --compose-file docker-compose.yml asdf`. The symptom: logstash starts up and immediately quits, presumably because it uses the default stdin config.

Running the same file with `docker-compose up` works. Running Logstash 5.x works.

The culprit MAY be that for some reason the container startup command (obtained by `docker ps --no-trunc`) for the working setups is `"/usr/local/bin/docker-entrypoint"` but for some reason the command for the broken 6.x container in the Docker Swarm is `"/docker-entrypoint -e "`. In logstash 6.x the empty `-e` flag overrides the configuration file (as far as I know), even if the file mounts correctly into the container.

I have no idea why this weird command is generated in the Swarm but not in compose. Setting the `entrypoint` property in the compose file to the "right" value makes logstash not start at all.

**Comment by jarpy** [comment]
Hi. Thanks for this really detailed report, and sorry for the slow response.

I'm sorry to say that this doesn't reproduce for me...

I created this `docker-compose.yml`
``` yaml
version: "3"

services:
  logstash:
      image: docker.elastic.co/logstash/logstash:6.2.4
      volumes:
        - ./logstash/pipeline/:/usr/share/logstash/pipeline/
      deploy:
        mode: replicated
        replicas: 1
```

...and this file in `./logstash/pipeline/pipeline.conf`:

```
input {
  heartbeat {
    message  => "I am alive."
    interval => 1
  }
}

output {
  stdout {}
}
```

I then ran `docker swarm init` and:

```
docker stack deploy --compose-file docker-compose.yml repro
```

After some time, I was able to see the exepected output in the logs:

```
$ docker service logs repro_logstash | grep alive | tail
repro_logstash.1.ygeqt11bnd7u@rasa    |        "message" => "I am alive.",
repro_logstash.1.ygeqt11bnd7u@rasa    |        "message" => "I am alive.",
repro_logstash.1.ygeqt11bnd7u@rasa    |        "message" => "I am alive.",
repro_logstash.1.ygeqt11bnd7u@rasa    |        "message" => "I am alive.",
repro_logstash.1.ygeqt11bnd7u@rasa    |        "message" => "I am alive.",
repro_logstash.1.ygeqt11bnd7u@rasa    |        "message" => "I am alive.",
repro_logstash.1.ygeqt11bnd7u@rasa    |        "message" => "I am alive.",
repro_logstash.1.ygeqt11bnd7u@rasa    |        "message" => "I am alive.",
repro_logstash.1.ygeqt11bnd7u@rasa    |        "message" => "I am alive.",
repro_logstash.1.ygeqt11bnd7u@rasa    |        "message" => "I am alive.",
```

Also, the command was set correctly:

```
$ docker ps --no-trunc --format '{{.Names}} {{.Command}}'
repro_logstash.1.ygeqt11bnd7urrawk5fe2xuev "/usr/local/bin/docker-entrypoint"
```

I suspect that Swarm is running the wrong image, especially since the [entrypoint/command for the old image from Docker Inc. is `/docker-entrypoint -e`](https://github.com/docker-library/logstash/blob/7339f3d967a5e725c8c40c8b68d0e035b6005064/5/Dockerfile#L68-L69
). How and why it's doing that are interesting questions!

**Comment by vlindhol** [comment]
I see! Thanks for the extra info. I'll try to debug some more. If I understand correctly, the `/docker-entrypoint -e` command worked in Docker <5.x because the `-e` flag was combined with any potential config files, while in 6.x it replaces the config file. Since this command now breaks my 6.2.4 image, I am assuming that the underlying software IS in fact 6.2.4, but the Dockerfile somehow behaves weirdly. Or perhaps Swarm, somehow.

**Comment by jarpy** [comment]
Your analysis sounds good to me. I can't find the documentation for the change to `-e` just now, but it does sound very familiar.

The most suspicious thing for me is still where the invalid command is coming from. It's also _super_ suspicious that it matches the command for a different Logstash image.

Good luck, and please let me know if you crack it! I'm very curious now.

**Comment by jarpy** [comment]
Oh, since you are mounting all of `/usr/share/logstash/config/`, make sure it's up to date with what's needed for 6.2. There are quite a few files that ship in that directory as the default configuration, including the new `pipelines.yml`.

```
$ docker run docker.elastic.co/logstash/logstash:6.2.4 ls /usr/share/logstash/config
jvm.options
log4j2.properties
logstash.yml
pipelines.yml
startup.options
```

**Comment by vlindhol** [comment]
Finally solved it! It was indeed Swarm running the wrong image. Turns out the example I gave above, which was slightly cleaned up from the production config I was using, was in itself working :) In actuality, I re-tagged the logstash image to shorten the image name, and that's where the problems arose. I have filed a new bug with Docker proper, at https://github.com/docker/cli/issues/1286, and this can be considered closed.
